woocommerce/woocommerce-gutenberg-products-block
# Fix E2E tests

This PR:
- fixes the E2E test suite.
- fixes the interactive mode (fixes #6608).
- add flaky tests report.


## E2E test suite

1. All the packages related to E2E tests are been updated. 
2. I parallelized the execution of the tests to improve the speed: for each E2E workflow (Gutenberg and without Gutenberg) 5 containers will be created which will perform different tests ([related code #1](https://github.com/woocommerce/woocommerce-blocks/pull/6639/files#diff-979aadf7fa69d93bf1a2bde6671ba35c54e1dec3e9afc63af46a460ac12d3c27R10-R13)), ([related code #2](https://github.com/woocommerce/woocommerce-blocks/pull/6639/files#diff-979aadf7fa69d93bf1a2bde6671ba35c54e1dec3e9afc63af46a460ac12d3c27R71-R72)).
3. Improve tests to reduce flakiness. 
4. Add a retry attempt (only for the CI). Now, if the test will fail, jest will execute the test again. If the test will fail twice, the CI will return an error, otherwise, the flaky reporter will add an issue in our repo. This is the same approach used by Gutenberg. Also, it seems that it same approach [used by Google ](https://testing.googleblog.com/2016/05/flaky-tests-at-google-and-how-we.html) and other companies.


#### Testing

1. Check that the CI doesn't return any error.

## Interactive mode
I fixed the interactive mode by adding a [dedicated jest-puppeteer config](https://github.com/woocommerce/woocommerce-blocks/pull/6639/files#diff-87c864ca36d6b2d84febe9066b00c93d152a09d3fecd9d2c3cad5a0516dcc8c3) for dev purpose.

#### Testing

1. Run commands `test:e2e-dev` and `test:e2e-dev-watch`. 
2. Check that `Chromium` is launched and you can see the test execution.

## Flaky tests report

For the flaky tests report, I adopted the same code that is used by the Gutenberg team. Now, if after a retry, the test passes, new artifacts that contain the flaky test data will be added. At the end of the CI, if the test suite passes, the flaky tests reporter will open an issue (or update the existing issue) for all flaky tests detected. 

#### Testing

The flaky tests reporter job uses the trigger `workflow_run`. To be triggered, the file must be already in the default branch. So you have to create a fork on your personal repo. After that:
1. Enable CI integration and issues from GitHub repo settings. Set the branch `fix/e2e-tests` as default branch
2. Create a PR from `fix/e2e-tests` to the `trunk` branch for triggering the CI.
3. Wait for CI.
4. Click on the job related to the E2E tests and click on `Summary` and check if there are visible artifacts called `flaky-test-report`.
![image](https://user-images.githubusercontent.com/4463174/177976411-22ea8452-21e4-4dd9-bf8a-4f5d5a74f0f3.png)
5. If yes, the flaky test reporter job should open an issue in the repo. You can find an example [on my personal fork](https://github.com/gigitux/woocommerce-blocks/issues/2).
